### PR TITLE
Fix test data over-memory bug

### LINKED ISSUE #756: GeoIPS full_install fails with recommended minimum memory
# Requested Update

### Description
*  The minimum recommended memory for GeoIPS is 12GB
* A 11.2GB file is downloaded by full test, and is loaded into system memory
* This causes the download to timeout and the system to stall

### Environment
*  Memory limited docker container

### Code to demonstrate issue
*  run test/integration_tests/full_install.sh

### PR DESCRIPTION
**Pull request by biosafetylvl5**
# Relevant XKCD

image

# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] NO REQUIRED ***unit tests*** (explain why not required)
* [x] Required ***integration tests*** added and pass for new/modified functionality
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** 

# Related Issues
fixes #756 
related to #660 
moves toward completing #758

# Testing Instructions

## Basic testing

0. Enter a 12GB RAM-constrained environment. (I am emulating via docker)
1. Run full_install
2. Run full_test

I'm passing all tests I'm passing on `main`.

## Test new helper script

``` bash
geoips$ python3 ./setup/download_test_data.py --help
Usage: download_test_data.py [-h] [--output-dir OUTPUT_DIR] [--no-rich] [--test-data-available]
                             [--list-test-datasets]
                             [input]

Download test data for GeoIPS

Positional Arguments:
  input                 The test data set to download, URL to the .tgz file or URL to git repo.

Options:
  -h, --help            show this help message and exit
  --output-dir OUTPUT_DIR
                        The directory to extract/clone files to.
  --no-rich             Disable rich text formatting and progress bars.
  --test-data-available
                        Returns 0 if available, else 1.
  --list-test-datasets  List test data sets available for download.
```

Just take it for a spin 🚗💨

# Summary

## Changes

0. Fixed memory bug by streaming the direct download and decompression instead of downloading into memory, writing to `stdout` and in to `tar` via a bash pipe.
1. Refactored the CIRA direct download urls into a yaml file and out of the existing `check_system_requirements.sh` file
2. Changed `download_test_data.py` to accept test data set names as well as urls
4. Added QOL functionality to `download_test_data.py`

## Rationale

Fixes bug.

Moves us closer to moving functionality into the GeoIPS CLI by moving some of the functionality from `check_system_requirements.sh` into python.

Makes adding/removing test data set URLs very easy.

Adds standalone tool for manually downloading test-data sets. (bonus, unsure how helpful)

### Notes

I use `rich`, a python module for nice CLI output. It is optional. Most GeoIPS users will have it installed as it is a sub-dependency of packages in the `lint` and `doc` extras of GeoIPS.

# Output

``` bash
geoips_user@5db51ce00289:/app/geoips_packages/geoips$ python3 ./setup/download_test_data.py --help
Usage: download_test_data.py [-h] [--output-dir OUTPUT_DIR] [--no-rich] [--list-test-datasets] [input]

Download test data for GeoIPS

Positional Arguments:
  input                 The test data set to download, URL to the .tgz file or URL to git repo.

Options:
  -h, --help            show this help message and exit
  --output-dir OUTPUT_DIR
                        The directory to extract/clone files to.
  --no-rich             Disable rich text formatting and progress bars.
  --list-test-datasets  List test data sets available for download.
```

## From `full_install`

``` bash
drwxr-xr-x 2 geoips_user geoips_user 40960 Aug 23 22:34 /app/geoips_packages/shapefiles/natural_earth/physical

Install log: /output/logs/20240823.223401_install.log
SUCCESS: source repo geoips appears to be installed successfully
    drwxr-xr-x 1 geoips_user geoips_user 4096 Aug 20 21:59 /app/geoips_packages/geoips

Install log: /output/logs/20240823.223401_install.log
Installing source repo data_fusion .... 
  /app/geoips_packages/data_fusion/
SUCCESS: Cloned and installed source repo data_fusion

Install log: /output/logs/20240823.223405_install.log
Installing source repo recenter_tc .... 
  /app/geoips_packages/recenter_tc/
SUCCESS: Cloned and installed source repo recenter_tc

Install log: /output/logs/20240823.223412_install.log
Installing source repo geoips_clavrx .... 
  /app/geoips_packages/geoips_clavrx/
SUCCESS: Cloned and installed source repo geoips_clavrx

Install log: /output/logs/20240823.223418_install.log
Installing source repo geoips_plugin_example .... 
  /app/geoips_packages/geoips_plugin_example/
SUCCESS: Cloned and installed source repo geoips_plugin_example

Install log: /output/logs/20240823.223420_install.log
Installing source repo template_basic_plugin .... 
  /app/geoips_packages/template_basic_plugin/
SUCCESS: Cloned and installed source repo template_basic_plugin

Install log: /output/logs/20240823.223423_install.log
Installing settings repo .vscode .... 
  /app/geoips_packages/.vscode/
SUCCESS: Cloned settings repo .vscode

Install log: /output/logs/20240823.223424_install.log
Installing tests data repo test_data_noaa_aws .... 
  /data/testdata/long-test-dir-name/test_data_noaa_aws/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_noaa_aws.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_noaa_aws @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_noaa_aws.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_noaa_aws.tgz to 
/data/testdata/long-test-dir-name...
File is 2.8 GiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_noaa_aws

Install log: /output/logs/20240823.223649_install.log
Installing tests data repo test_data_amsr2 .... 
  /data/testdata/long-test-dir-name/test_data_amsr2/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_amsr2_1.6.0.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_amsr2 @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_amsr2_1.6.0.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_amsr2_1.6.0.tgz to 
/data/testdata/long-test-dir-name...
File is 384.9 MiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_amsr2

Install log: /output/logs/20240823.223722_install.log
Installing tests data repo test_data_clavrx .... 
  /data/testdata/long-test-dir-name/test_data_clavrx/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_clavrx_1.10.0.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_clavrx @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_clavrx_1.10.0.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_clavrx_1.10.0.tgz to 
/data/testdata/long-test-dir-name...
File is 2.9 GiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_clavrx

Install log: /output/logs/20240823.223924_install.log
Installing tests data repo test_data_gpm .... 
  /data/testdata/long-test-dir-name/test_data_gpm/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_gpm_1.6.0.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_gpm @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_gpm_1.6.0.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_gpm_1.6.0.tgz to 
/data/testdata/long-test-dir-name...
File is 1.1 GiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_gpm

Install log: /output/logs/20240823.224010_install.log
Installing tests data repo test_data_sar .... 
  /data/testdata/long-test-dir-name/test_data_sar/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_sar_1.12.2.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_sar @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_sar_1.12.2.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_sar_1.12.2.tgz to 
/data/testdata/long-test-dir-name...
File is 112.0 MiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_sar

Install log: /output/logs/20240823.224029_install.log
Installing tests data repo test_data_scat .... 
  /data/testdata/long-test-dir-name/test_data_scat/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_scat_1.11.3.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_scat @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_scat_1.11.3.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_scat_1.11.3.tgz to 
/data/testdata/long-test-dir-name...
File is 358.5 MiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_scat

Install log: /output/logs/20240823.224045_install.log
Installing tests data repo test_data_smap .... 
  /data/testdata/long-test-dir-name/test_data_smap/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_smap_1.6.0.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_smap @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_smap_1.6.0.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_smap_1.6.0.tgz to 
/data/testdata/long-test-dir-name...
File is 172.4 MiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_smap

Install log: /output/logs/20240823.224108_install.log
Installing tests data repo test_data_viirs .... 
  /data/testdata/long-test-dir-name/test_data_viirs/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_viirs_1.6.0.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_viirs @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_viirs_1.6.0.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_viirs_1.6.0.tgz to 
/data/testdata/long-test-dir-name...
File is 11.2 GiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_viirs

Install log: /output/logs/20240823.224907_install.log
Installing tests data repo test_data_fusion .... 
  /data/testdata/long-test-dir-name/test_data_fusion/ from https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_fusion.tgz via cira
DOWNLOADING: NextCloud Dataset test_data_fusion @ https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_fusion.tgz
Downloading and extracting 
https://io2.cira.colostate.edu/s/J73tEcn22smktMi/download?path=%2F&files=test_data_fusion.tgz to 
/data/testdata/long-test-dir-name...
File is 2.7 GiB... 
Success. Files downloaded and extracted.
SUCCESS: Decompressed test_data_fusion
23_225053 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/geoips_plugin_example/geoips_plugin_example/registered_plugins.json
23_225053 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/template_basic_plugin/my_package/registered_plugins.json
23_225054 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/data_fusion/data_fusion/registered_plugins.json
23_225054 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/recenter_tc/recenter_tc/registered_plugins.json
23_225054 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/geoips_clavrx/geoips_clavrx/registered_plugins.json
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'tc_file_naming' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'remss_reader' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'geostationary_geolocation' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'hrit_reader' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'interp_pyresample' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'interp_scipy' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:820  INTERACTIVE: Skipping module 'boxdefinitions' from 'geoips', interface_name is 'None'
23_225054 create_plugin_registries.py:318  INTERACTIVE: Writing /app/geoips_packages/geoips/geoips/registered_plugins.json
```

### CODE REVIEW

File: docs/source/releases/latest/better-download-test-data.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+bug fix:
+- description: |
```
Comment (mindyls) [inline docs/source/releases/latest/better-download-test-data.yaml:2]: Note: don't merge this until 1.14.0 release is complete

File: setup/check_system_requirements.sh [hunk 2]
```diff
@@ -356,9 +338,9 @@ if [[ "$1" == "test_data" || "$1" == "test_data_github" ]]; then
             exit 1
         fi
         echo "Installing $test_data_name_string .... "
-        echo "  $test_data_dir/ from $test_data_url via $test_data_source_location"
+        echo "  $test_data_dir/ using $test_data_url via $test_data_source_location"
         if [[ "$test_data_source_location" == "github" ]]; then
-            python $SCRIPT_DIR/download_test_data.py $test_data_url $test_data_dir >> $install_log 2>&1
+            python3 $SCRIPT_DIR/download_test_data.py $test_data_url --output-dir $GEOIPS_TESTDATA_DIR >> $install_log 2>&1
```
Comment (mindyls) [inline setup/check_system_requirements.sh:343]: Is test_data_dir always GEOIPS_TESTDATA_DIR/reponame? Should we remove the test data dir variable for consistency? Is test data dir ever something besides geoips test data dir?

**Review by mindyls — DISMISSED** [review]
Thank you for making rich optional! We can merge after release finished

File: setup/check_system_requirements.sh (hunk 2 above)
Comment (biosafetylvl5) [reply setup/check_system_requirements.sh:343]: Yes, it's set like this in the original script `test_data_dir=$GEOIPS_TESTDATA_DIR/$test_data_name`. It's currently still used by the git clones. We could remove it and download the git repos with the script (which already has the ability to do so)

Comment (mindyls) [reply setup/check_system_requirements.sh:343]: Isn't it already downloading git repos with the download script?  If yes and test_data_dir is no longer used,  I say remove test data dir variable. If it's still used,  just resolve this conversation,  we can deal with it later,  if we need to.

Comment (biosafetylvl5) [reply setup/check_system_requirements.sh:343]: Sorry - yes, you're right. I forgot I did that. Unfortunately still used in a few other small places. I say we deal with later if we need to

File: docs/source/releases/latest/better-download-test-data.yaml (hunk 1 above)
Comment (biosafetylvl5) [reply docs/source/releases/latest/better-download-test-data.yaml:2]: switched to dev-staging, so GTG

**Review by jsolbrig — APPROVED** [review]

**Review by mindyls — APPROVED** [review]
No additional changes,  base branch changed to dev staging.  Reapproving

**Pull request merged**